woocommerce/woocommerce
# Filter the cart item quantity validation in StoreApi

### Changes proposed in this Pull Request:

Fixes woocommerce/woocommerce#52263.

Filter the cart item quantity validation via Shop API to make it possible for third party plugins to prevent changing cart item quantity after checking their on their own conditions.

**Why it is necessary**<br>Third party developers cannot **validate a cart item quantity change with their custom conditions** with the storeAPI.<br>This used to be possible with the old API either with: “woocommerce_stock_amount_cart_item” or “woocommerce_update_cart_validation” hooks.

**What does it change**<br>This PR simply adds a PHP hook ([apply_filters](<https://developer.wordpress.org/reference/functions/apply_filters/>)) to the [Automattic\\WooCommerce\\StoreApi\\Utilities\\QuantityLimits validate_cart_item_quantity()](<https://github.com/woocommerce/woocommerce/blob/9.4.1/plugins/woocommerce/src/StoreApi/Utilities/QuantityLimits.php#L130>) method return value to allow third party developers to prevent changing a cart item quantity and to display an error message explaining the reason.

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](<https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions>), include your detailed testing instructions:

In this example, we are a third-party developer who wants to prevent the cart item quantity from being greater than 3.

1. Create a test plugin at `wp-content/plugins/test/test.php`
2. It should content the following code:

```
<?php
/**
 * Plugin Name: Test woocommerce_store_api_validate_cart_item_quantity hook
 * Requires Plugins: woocommerce
 */

function test_wc_store_api_validate_cart_item_quantity( $true, $quantity, $cart_item ) {
	$custom_condition = $quantity <= 3;
	if( ! $custom_condition ) {
		return new \WP_Error(
			'invalid_custom_condition',
			'The quantity added to the cart doesn\'t meet the custom condition'
		);
	}
	
	return $true;
}
add_filter( 'woocommerce_store_api_validate_cart_item_quantity', 'test_wc_store_api_validate_cart_item_quantity', 10, 3 );
```

3. From your site admin, go to Plugins > All plugins > Make sure WooCommerce is active, deactivate everything else, and activate this test plugin
4. Make sure you are using the Block Cart (as opposed to the Classic Cart)
5. Create a product
6. Add this product to cart
7. Change the cart item quantity to 3, it should work
8. Change the cart item quantity to 4, an error message should be displayed
9. Refresh the page, the quantity should be reverted

### Changelog entry


+++

#### Significance


#### Type


#### Message

Filters the quantity validation for a cart item being updated via the Store API.

+++
